mava/systems/ff_ippo_lbf.py

```diff
@@ -0,0 +1,638 @@
+        * config["arch"]["num_envs"]
+    )
+    # Get total_timesteps
+    config["system"]["total_timesteps"] = (
+        n_devices
+        * config["system"]["num_updates"]
+        * config["system"]["rollout_length"]
+        * config["system"]["update_batch_size"]
+        * config["arch"]["num_envs"]
+    )
+    pprint(config)
+
+    # Run experiment for a total number of evaluations.
+    max_episode_return = jnp.float32(0.0)
+    best_params = None
+    for i in range(config["arch"]["num_evaluation"]):
+        # Train.
+        start_time = time.time()
+
+        learner_output = learn(learner_state)
+        jax.block_until_ready(learner_output)
+
+        # Log the results of the training.
+        elapsed_time = time.time() - start_time
+        learner_output.episodes_info["steps_per_second"] = steps_per_rollout / elapsed_time
+        log(
+            metrics=learner_output,
+            t_env=steps_per_rollout * (i + 1),
+            trainer_metric=True,
+        )
+
+        # Prepare for evaluation.
+        start_time = time.time()
+        trained_params = jax.tree_util.tree_map(
+            lambda x: x[:, 0, ...],
+            learner_output.learner_state.params.actor_params,  # Select only actor params
+        )
+        rng_e, *eval_rngs = jax.random.split(rng_e, n_devices + 1)
+        eval_rngs = jnp.stack(eval_rngs)
+        eval_rngs = eval_rngs.reshape(n_devices, -1)
+
+        # Evaluate.
+        evaluator_output = evaluator(trained_params, eval_rngs)
+        jax.block_until_ready(evaluator_output)
+
+        # Log the results of the evaluation.
+        elapsed_time = time.time() - start_time
+        evaluator_output.episodes_info["steps_per_second"] = steps_per_rollout / elapsed_time
+        episode_return = log(
+            metrics=evaluator_output,
+            t_env=steps_per_rollout * (i + 1),
+        )
+        if config["arch"]["absolute_metric"] and max_episode_return <= episode_return:
+            best_params = copy.deepcopy(trained_params)
+            max_episode_return = episode_return
+
+        # Update runner state to continue training.
+        learner_state = learner_output.learner_state
+
+    # Measure absolute metric.
+    if config["arch"]["absolute_metric"]:
+        start_time = time.time()
+
+        rng_e, *eval_rngs = jax.random.split(rng_e, n_devices + 1)
+        eval_rngs = jnp.stack(eval_rngs)
+        eval_rngs = eval_rngs.reshape(n_devices, -1)
+
+        evaluator_output = absolute_metric_evaluator(best_params, eval_rngs)
+        jax.block_until_ready(evaluator_output)
+
+        elapsed_time = time.time() - start_time
+        evaluator_output.episodes_info["steps_per_second"] = steps_per_rollout / elapsed_time
+        log(
+            metrics=evaluator_output,
+            t_env=steps_per_rollout * (i + 1),
+            absolute_metric=True,
+        )
+
+
+@hydra.main(config_path="../configs", config_name="default_ff_ippo.yaml", version_base="1.2")
```

Comment:
It will also be good if we have a default_ff_ippo_lbf config or something along those lines? Because this will probably not do what we want if the default ppo config points to rware.

Reply:
I thought that we simply make changes in the default files but if this makes it easier to run experiments later I'll be happy to quickly add them!